### PR TITLE
Faster bencoding

### PR DESCRIPTION
**Pull request by alanmcgovern**

### CODE REVIEW

File: src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        {
+            BEncodedString key = null;
+            BEncodedValue value = null;
+            var torrent = new BEncodedDictionary ();
+            if (reader.ReadByte () != 'd')
+                throw new BEncodingException ("Invalid data found. Aborting"); // Remove the leading 'd'
+
+            int read;
+            while ((read = reader.ReadByte ()) != -1 && read != 'e') {
+                key = (BEncodedString) Decode (reader);         // keys have to be BEncoded strings
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs]: This looks like a bug. We're discarding the previous read.

File: src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+                read = reader.ReadByte ();
```
Comment (alanmcgovern) [inline src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs]: Similar issue here?

Comment (alanmcgovern) [reply src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs]: Tests FTW!

File: src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs (hunk 1 above)
Comment (alanmcgovern) [reply src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs]: Tests FTW!

File: src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+            while ((read = reader.ReadByte ()) != -1 && read != ':') {
+                if (read < '0' || read > '9')
+                    throw new BEncodingException ($"Invalid BEncodedString. Length was '{length}' instead of a number");
+                length = length * 10 + (read - '0');
+            }
+
+            if (read != ':')
+                throw new BEncodingException ("Invalid data found. Aborting");
+
+            var bytes = new byte[length];
```
Comment (Therzok) [inline src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs]: Are these strings really long? Otherwise maybe ArrayPool<byte> could be used here.

Comment (alanmcgovern) [reply src/MonoTorrent/MonoTorrent.BEncoding/BEncodeDecoder.cs]: The byte[] is retained in the BEncodedString object forever. A BEncodedString is a raw byte[] which can sometimes be converted into a valid UTF-8 string. If a UTF-8 string is needed it is generated on the fly and not retained

**Pull request merged**